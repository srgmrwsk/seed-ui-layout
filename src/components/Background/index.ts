export * from './Background';
export { default as Background } from './Background';
