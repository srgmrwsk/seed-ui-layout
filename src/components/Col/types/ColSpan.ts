export type ColSpan = boolean | number | 'auto';
