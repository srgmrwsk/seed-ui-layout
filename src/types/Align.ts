export type Align = 'start' | 'end' | 'center' | 'baseline' | 'stretch';
