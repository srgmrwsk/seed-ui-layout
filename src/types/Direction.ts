export type Direction = 'column' | 'column-reverse' | 'row' | 'rew-reverse';
