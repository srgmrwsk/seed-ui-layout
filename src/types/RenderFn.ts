import React from 'react';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type RenderFn = (component: React.ReactElement, ...args: []) => any;
