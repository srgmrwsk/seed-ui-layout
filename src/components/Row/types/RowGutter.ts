import { Sizing } from '../../../types/Sizing';

export type RowGutter = Sizing | [Sizing, Sizing];
