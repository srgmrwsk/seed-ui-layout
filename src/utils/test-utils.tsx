import React from 'react';
import { render } from '@testing-library/react';
import { renderToStaticMarkup } from 'react-dom/server';
import { configureAxe } from 'jest-axe';

import { RenderFn } from '../types/RenderFn';

export const axe = configureAxe({
  rules: {
    // disable landmark rules when testing isolated components
    region: { enabled: false },
  },
});

export function create(
  ...args: Parameters<RenderFn>
): ChildNode | HTMLCollection | null {
  const { container } = render(...args);
  return container.children.length > 1
    ? container.children
    : container.firstChild;
}
export function renderToHtml(component: React.ReactElement): string {
  return renderToStaticMarkup(component);
}
