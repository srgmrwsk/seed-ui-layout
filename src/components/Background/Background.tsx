import React from 'react';
import cx from 'classnames';

import { ResponsiveValue } from '../../types/ResponsiveValue';
import {
  BackgroundPosition,
  getBgPosVariables,
} from '../../utils/get-bg-pos-variables';
import {
  BackgroundRepeat,
  getBgRepeatVariables,
} from '../../utils/get-bg-repeat-variables';
import {
  BackgroundSize,
  getBgSizeVariables,
} from '../../utils/get-bg-size-variables';
import { getBgImageVariables } from '../../utils/get-bg-image-variables';
import { getBooleanStyles } from '../../utils/get-boolean-styles';

export interface BackgroundProps
  extends Omit<React.AllHTMLAttributes<HTMLElement>, 'as'> {
  as?: React.ElementType;
  backgroundImage?: ResponsiveValue<string>;
  backgroundPosition?: ResponsiveValue<BackgroundPosition>;
  backgroundRepeat?: ResponsiveValue<BackgroundRepeat>;
  backgroundSize?: ResponsiveValue<BackgroundSize>;
  children?: React.ReactNode;
}

function Background(
  {
    as: As = 'div',
    backgroundImage,
    backgroundPosition,
    backgroundRepeat,
    backgroundSize,
    children,
    style,
    className,
    ...elementProps
  }: BackgroundProps,
  ref: React.Ref<HTMLDivElement>,
): JSX.Element {
  return (
    <As
      ref={ref}
      className={cx(
        getBooleanStyles('_seed-bg-image', backgroundImage),
        getBooleanStyles('_seed-bg-pos', backgroundPosition),
        getBooleanStyles('_seed-bg-repeat', backgroundRepeat),
        getBooleanStyles('_seed-bg-size', backgroundSize),
        className,
      )}
      style={{
        ...getBgImageVariables('seed-bg-image', backgroundImage),
        ...getBgPosVariables('seed-bg-pos', backgroundPosition),
        ...getBgRepeatVariables('seed-bg-repeat', backgroundRepeat),
        ...getBgSizeVariables('seed-bg-size', backgroundSize),
        ...style,
      }}
      {...elementProps}
    >
      {children}
    </As>
  );
}

Background.displayName = 'Background';

export default React.forwardRef(Background);
