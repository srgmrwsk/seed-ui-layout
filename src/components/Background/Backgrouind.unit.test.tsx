import React from 'react';
import { renderToStaticMarkup } from 'react-dom/server';

import { RenderFn } from '../../types/RenderFn';
import { axe, create } from '../../utils/test-utils';

import Background, { BackgroundProps } from './Background';

describe('Background', () => {
  const setup = (renderFn: RenderFn, props?: BackgroundProps) =>
    renderFn(<Background {...props} />);

  it('should render correctly with default props', () => {
    const actual = setup(create);
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with as prop', () => {
    const actual = setup(create, { as: 'section' });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with children prop', () => {
    const actual = setup(create, { children: <p>Element with text.</p> });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with background-image prop', () => {
    const actual = setup(create, { backgroundImage: '/images/image.jpg' });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with background-image responsive prop', () => {
    const actual = setup(create, {
      backgroundImage: {
        xs: '/images/image.jpg',
        sm: '/images/image-sm.jpg',
        md: '/images/image-md.jpg',
        lg: '/images/image-lg.jpg',
        xl: '/images/image-xl.jpg',
        xxl: '/images/image-xxl.jpg',
      },
    });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with background-position prop', () => {
    const actual = setup(create, { backgroundPosition: 'center' });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with background-position responsive prop', () => {
    const actual = setup(create, {
      backgroundPosition: {
        xs: 'left',
        sm: 0.5,
        md: 50,
        lg: [0.5, 50],
        xl: ['left', 50],
        xxl: ['left', 'top'],
      },
    });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with background-size prop', () => {
    const actual = setup(create, { backgroundSize: 'cover' });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with background-size responsive prop', () => {
    const actual = setup(create, {
      backgroundSize: {
        xs: 'cover',
        sm: 'contain',
        md: [250, 375],
        lg: [250, 0.375],
        xl: [0.25, 375],
        xxl: [250, '75vh'],
      },
    });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with background-repeat prop', () => {
    const actual = setup(create, { backgroundRepeat: 'no-repeat' });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with background-repeat responsive prop', () => {
    const actual = setup(create, {
      backgroundRepeat: {
        xs: 'no-repeat',
        sm: 'repeat',
        md: 'repeat-x',
        lg: 'repeat-y',
        xl: 'round',
        xxl: ['space', 'no-repeat'],
      },
    });
    expect(actual).toMatchSnapshot();
  });

  it('should meet accessibility guidelines', async () => {
    const wrapper = setup(renderToStaticMarkup);
    const actual = await axe(wrapper);
    expect(actual).toHaveNoViolations();
  });
});
