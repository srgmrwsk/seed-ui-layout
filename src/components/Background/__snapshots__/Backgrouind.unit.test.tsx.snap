// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Background should render correctly with as prop 1`] = `
<section
  class=""
/>
`;

exports[`Background should render correctly with background-image prop 1`] = `
<div
  class="_seed-bg-image"
  style="--seed-bg-image: /images/image.jpg;"
/>
`;

exports[`Background should render correctly with background-image responsive prop 1`] = `
<div
  class="_seed-bg-image _seed-bg-image-sm _seed-bg-image-md _seed-bg-image-lg _seed-bg-image-xl _seed-bg-image-xxl"
  style="--seed-bg-image: /images/image.jpg; --seed-bg-image-sm: /images/image-sm.jpg; --seed-bg-image-md: /images/image-md.jpg; --seed-bg-image-lg: /images/image-lg.jpg; --seed-bg-image-xl: /images/image-xl.jpg; --seed-bg-image-xxl: /images/image-xxl.jpg;"
/>
`;

exports[`Background should render correctly with background-position prop 1`] = `
<div
  class="_seed-bg-pos"
  style="--seed-bg-pos: center;"
/>
`;

exports[`Background should render correctly with background-position responsive prop 1`] = `
<div
  class="_seed-bg-pos _seed-bg-pos-sm _seed-bg-pos-md _seed-bg-pos-lg _seed-bg-pos-xl _seed-bg-pos-xxl"
  style="--seed-bg-pos: left; --seed-bg-pos-sm: 50%; --seed-bg-pos-md: 50px; --seed-bg-pos-lg: 50% 50px; --seed-bg-pos-xl: left 50px; --seed-bg-pos-xxl: left top;"
/>
`;

exports[`Background should render correctly with background-repeat prop 1`] = `
<div
  class="_seed-bg-repeat"
  style="--seed-bg-repeat: no-repeat;"
/>
`;

exports[`Background should render correctly with background-repeat responsive prop 1`] = `
<div
  class="_seed-bg-repeat _seed-bg-repeat-sm _seed-bg-repeat-md _seed-bg-repeat-lg _seed-bg-repeat-xl _seed-bg-repeat-xxl"
  style="--seed-bg-repeat: no-repeat; --seed-bg-repeat-sm: repeat; --seed-bg-repeat-md: repeat-x; --seed-bg-repeat-lg: repeat-y; --seed-bg-repeat-xl: round; --seed-bg-repeat-xxl: space no-repeat;"
/>
`;

exports[`Background should render correctly with background-size prop 1`] = `
<div
  class="_seed-bg-size"
  style="--seed-bg-size: cover;"
/>
`;

exports[`Background should render correctly with background-size responsive prop 1`] = `
<div
  class="_seed-bg-size _seed-bg-size-sm _seed-bg-size-md _seed-bg-size-lg _seed-bg-size-xl _seed-bg-size-xxl"
  style="--seed-bg-size: cover; --seed-bg-size-sm: contain; --seed-bg-size-md: 250px 375px; --seed-bg-size-lg: 250px 37.5%; --seed-bg-size-xl: 25% 375px; --seed-bg-size-xxl: 250px 75vh;"
/>
`;

exports[`Background should render correctly with children prop 1`] = `
<div
  class=""
>
  <p>
    Element with text.
  </p>
</div>
`;

exports[`Background should render correctly with default props 1`] = `
<div
  class=""
/>
`;
